File: SINGLYCL.c
/////////////////////////////////////////////////////////////////////////////////
// Application Name: SINGLYCL
// Description: It is a Singly Circular linked list (Data structure) application.
// Date: 21/09/2021.
// Author Name: Harish Vijay Bavne.
/////////////////////////////////////////////////////////////////////////////////

// Required Header files inclusion.
#include<stdio.h>
#include<stdlib.h>

// Creation of Node.
struct node
{
	int Data;
	struct node* Next;
};

// Required typedef's
typedef struct node NODE;
typedef struct node* PNODE;
typedef struct node** PPNODE;

// Function for insert node at first.
void InsertFirst(PPNODE Head,PPNODE Tail,int No)
{
	PNODE Newn = (PNODE)malloc(sizeof(NODE));
	Newn->Data = No;
	Newn->Next = NULL;
	
	if (*Head == NULL)
	{
		*Head = Newn;
		*Tail = Newn;
	}
	else
	{
		Newn->Next = *Head;
		*Head = Newn;
	}
	(*Tail)->Next = *Head;
}

// Function for insert node at last.
void InsertLast(PPNODE Head,PPNODE Tail,int No)
{
	PNODE Newn = (PNODE)malloc(sizeof(NODE));
	Newn->Data = No;
	Newn->Next = NULL;
	PNODE temp = *Head;
	if (*Head == NULL)
	{
		*Head = Newn;
		*Tail = Newn;
		(*Tail)->Next = *Head;
	}
	else
	{
		while(temp->Next != (*Head))
		{
			temp = temp->Next;
		}
		temp->Next = Newn;
		(*Tail) = Newn;
		(*Tail)->Next = *Head;
			
	}
}

// Function for display element's in Linked List.
void Display(PNODE Head)
{
	PNODE temp = Head;
	if (Head == NULL)
	{
		printf("Linked List is Empty.\n");
		return;
	}
	while(temp->Next != Head)
	{
		printf("|%d|->",temp->Data);
		temp = temp->Next;
	}
	printf("|%d|->",temp->Data);
	printf("|First|\n");
}

// Function for counting number of nodes present in Linked List.
int Count(PNODE Head)
{
	int iCnt = 0;
	if (Head == NULL)
	{
		return 0;
	}
	PNODE temp = Head;
	while(temp->Next != Head)
	{
		iCnt ++;
		temp = temp->Next;
	}
	iCnt ++;
	return iCnt;
}

// Function for inserting node at said position.
void InsertAtPosition(PPNODE Head,PPNODE Tail,int No,int Pos)
{
	PNODE Newn = (PNODE)malloc(sizeof(NODE));
	Newn->Data = No;
	Newn->Next = NULL;
	PNODE temp = *Head;
	int i = 0;
	int iSize = Count(*Head);
	
	if ((Pos < 1)||(Pos>iSize+1))
	{
		return;
	}
	else if(Pos == 1)
	{
		InsertFirst(Head,Tail,No);
	}
	else if(Pos == iSize+1)
	{
		InsertLast(Head,Tail,No);
	}
	else
	{
		for (i = 1;i<Pos-1;i++)
		{
			temp = temp->Next;
		}
		Newn->Next = temp->Next;
		temp->Next = Newn;
	}
}

// Function for delete first node.
void DeleteFirst(PPNODE Head,PPNODE Tail)
{
	PNODE temp = *Head;
	if (*Head == NULL)
	{
		return;
	}
	else if (temp->Next == (*Head))
	{
		free(temp);
		*Head = NULL;
		*Tail = NULL;
	}
	else
	{
		*Head = temp->Next;
		free(temp);
		(*Tail)-> Next= *Head;
	}
}

// Function for delete last node.
void DeleteLast(PPNODE Head,PPNODE Tail)
{
	PNODE temp = *Head;
	if (*Head == NULL)
	{
		return;
	}
	else if (temp->Next == (*Head))
	{
		free(temp);
		*Head = NULL;
		*Tail = NULL;
	}
	else
	{
		while(temp->Next->Next != *Head)
		{
			temp = temp->Next;
		}
		free(temp->Next);
		*Tail = temp;
		(*Tail)->Next = *Head;
	}
}

// Function for delete node at said position.
void DeleteAtPosition(PPNODE Head,PPNODE Tail,int Pos)
{
	int i = 0;
	int iSize = Count((*Head));
	PNODE temp = *Head;
	PNODE Target = NULL;
	if ((Pos < 1)||(Pos >iSize))
	{
		return;
	}
	else if(Pos == 1)
	{
		DeleteFirst(Head,Tail);
	}
	else if(Pos == iSize)
	{
		DeleteLast(Head,Tail);
	}
	else
	{
		for (i = 1;i<Pos-1;i++)
		{
			temp = temp->Next;
		}
		Target = temp->Next;
		temp->Next = Target->Next;
		free(Target);
	}
}

// Entry point function.
int main()
{
	int iRet = 0,choice = 1,Value = 0,Pos = 0;
	PNODE First = NULL;
	PNODE Tail = NULL;
	while (choice != 0)
	{
		printf("\n-----------------------------------------------------\n");
		printf("\n* Application of Singly Circular Linked List *\n\n");
		printf("1:Insert node at first\n");
		printf("2:Insert node at last\n");
		printf("3:Insert node at position\n");
		printf("4:Delete node at first\n");
		printf("5:Delete node at last\n");
		printf("6:Delete node at position\n");
		printf("7:Dispay node elements\n");
		printf("8:Count Number of nodes\n");
		printf("9:Exit the application\n");
		printf("\n-----------------------------------------------------\n");
		printf("Enter a choice:");
		scanf("%d",&choice);
		
		switch (choice)
		{
			case 1:
				printf("Enter a element:");
				scanf("%d",&Value);
				InsertFirst(&First,&Tail,Value);
				break;
			case 2:
				printf("Enter a element:");
				scanf("%d",&Value);
				InsertLast(&First,&Tail,Value);
				break;
			case 3:
				printf("Enter a element:");
				scanf("%d",&Value);
				
				printf("Enter a Position:");
				scanf("%d",&Pos);
				InsertAtPosition(&First,&Tail,Value,Pos);
				break;
			case 4:
				DeleteFirst(&First,&Tail);
				break;
			case 5:
				DeleteLast(&First,&Tail);
				break;
			case 6:
				printf("Enter a position:");
				scanf("%d",&Pos);
				DeleteAtPosition(&First,&Tail,Pos);
				break;
			case 7:
				Display(First);
				break;	
			case 8:
				iRet = Count(First);
				printf("Number of nodes are:%d\n",iRet);
				break;	
			case 9:
				printf("Thank you for using Apllication.");
				return 0;
			default:
				printf("Enter a correct choice.\n");
				break;	
		}
	}
	return 0;
}

File: SINGLYCL.py
##############################################################################
# Application Name: SINGLYCL
# Description: It is a Singly Circular linked list (Data structure) application.
# Date: 05/10/2021.
# Author Name: Harish Vijay Bavne
##############################################################################

# Node Declaration
class node:
    def __init__(self,no):
        self.Data = no
        self.next = None
        
# Class for Singly Circular Linked list
class SinglyCL:
    # class variable private for node count
    __Size = 0                      
    
    def __init__(self):
        # Head is for first node reference
        self.Head = None
        
        # Tail is for last node reference
        self.Tail = None
       
    # Function for Insert elements in linked list at first position
    def InsertFirst(self,No):       
        newn = node(No)
        if (self.Head == None):
            self.Head = newn
            self.Tail = newn
            self.Tail.next = self.Head
            SinglyCL.__Size = SinglyCL.__Size + 1
        else:
            newn.next = self.Head
            self.Head = newn
            self.Tail.next = self.Head
            SinglyCL.__Size = SinglyCL.__Size + 1
    
    # Function for Insert element in linked list at last position
    def InsertLast(self,No):
        newn = node(No)
        if (self.Head == None):
            self.Head = newn
            self.Tail = newn
            self.Tail.next = self.Head
            SinglyCL.__Size = SinglyCL.__Size + 1
        else:
            self.Tail.next = newn
            self.Tail = newn
            self.Tail.next = self.Head
            SinglyCL.__Size = SinglyCL.__Size + 1
        
    # Function for Insert element in linked list at said position
    def InsertAtPosition(self,No,Pos):
        newn = node(No)
        temp = self.Head
        iSize = SinglyCL.__Size
        if ((Pos < 1) or (Pos > iSize+1)):
            return
        if (Pos == 1):
            self.InsertFirst(No)
        elif (Pos == iSize+1):
            self.InsertLast(No)
        else:    
            print("P")
            for i in range(1,Pos-1):
                temp = temp.next
            newn.next = temp.next
            temp.next = newn
            SinglyCL.__Size = SinglyCL.__Size + 1
    
    # Function for Delete element in Linked list at first position
    def DeleteFirst(self):
        temp = self.Head
        if (self.Head == None):
            return
        elif (self.Head.next == None):
            del(temp)
            self.Head = None
            self.Tail = None
            SinglyCL.__Size = SinglyCL.__Size - 1
        else:
            self.Head = temp.next
            del(temp)
            self.Tail.next = self.Head
            SinglyCL.__Size = SinglyCL.__Size - 1
    
    # Function for Delete element in Linked list at Last position
    def DeleteLast(self):
        temp = self.Head
        if (self.Head == None):
            return
        elif (self.Head.next == None):
            del(temp)
            self.Head = None
            self.Tail = None
            SinglyCL.__Size = SinglyCL.__Size - 1
        else:
            for i in range(1,SinglyCL.__Size-1):
                temp = temp.next
            del(temp.next)
            self.Tail = temp
            self.Tail.next = self.Head
            SinglyCL.__Size = SinglyCL.__Size - 1
    
    # Function for Delete element at said position in Linked list
    def DeleteAtPosition(self,Pos):
        temp =self.Head
        iSize = SinglyCL.__Size
        target = None
        if ((Pos < 1)or(Pos > iSize)):
            return
        elif (Pos == 1):
            self.DeleteFirst()
        elif (Pos == iSize):
            self.DeleteLast()
        else:
            for i in range(1,Pos-1):
                temp = temp.next
            target = temp.next
            temp.next = target.next
            del(target)
            SinglyCL.__Size = SinglyCL.__Size - 1
    
    # Function for Display elements in linked list
    def Display(self):
        temp = self.Head
        for i in range(SinglyCL.__Size):
            print(f"|{temp.Data}|=>",end="")
            temp = temp.next
        print("Head")
        
    # Function for count number of Nodes in linked list
    def Count(self):
        return SinglyCL.__Size
    
# Entry point function main.
def main():
    obj = SinglyCL()
    def menu():
        print("*********************************")
        print("   Singly Circular Linked List   ")
        print("1:Insert First")
        print("2:Insert last")
        print("3:Insert at Position")
        print("4:DeleteFirst ")
        print("5:DeleteLast")
        print("6:DeleteAtPosition")
        print("7:Display element")
        print("8:Count Number of Nodes")
        print("0:Exit application")
        print("*********************************")
    menu()
    choice = int(input("Enter a your choice:"))
    while True:
        if (choice == 1):
            No = int(input("Enter Element:"))
            obj.InsertFirst(No)
        elif (choice == 2):
            No = int(input("Enter Element:"))
            obj.InsertLast(No)
        elif (choice == 3):
            No = int(input("Enter Element:"))
            Pos = int(input("Enter position:"))
            obj.InsertAtPosition(No,Pos)
        elif (choice == 4):
            obj.DeleteFirst()
        elif (choice == 5):
            obj.DeleteLast()
        elif (choice == 6):
            Pos = int(input("Enter a position:"))
            obj.DeleteAtPosition(Pos)
        elif (choice == 7):
            obj.Display()
        elif (choice == 8):
            Ret = obj.Count()
            print(f"Number of Nodes are:{Ret}")
        elif (choice == 0):
            print("Thank you for using application")
            break
        else:
            print("Enter a correct choice.")
        menu()
        choice = int(input("Enter a your choice:"))
        
# Starter
if __name__ == "__main__":
    main()